### PR TITLE
Upgrade to PG12

### PR DESCRIPTION
**Pull request by rdunklau**
Hello,

We have ported libpg_query to PG12 (see here: https://github.com/lfittl/libpg_query/pull/62) and this is an updated version of pglast to adapt it to the changes.

The main difference comes from the way we built libpg_query: instead of parsing the postgres source code to generate a stripped down version of it (which depends heavily on the clang version used to perform the introspection), we rely on the compiler / linker LTO to remove code we don't need. It should make it much easier to port libpg_query to later version of Postgres as they are released.

Unfortunately, this means that (at least at this time) we only support GCC and not clang.

What do you think about it ?

### CODE REVIEW

**Comment by lelit** [comment]
Wow, this is great, thank you!

I won't be able to dedicate required time to inspect the details in the near future, but I'll try to focus on it next week or so.

Did you receive any feedback by Lukas?

**Comment by rdunklau** [comment]
No feedback as of yet. 

I see that your CI is failing on this, I'll take a look at what is required to make it work. Probably a different compiler or linker ?

No problem, we can discuss this whenever you get time to look into it !

**Comment by rdunklau** [comment]
I don't understand how the CI works, but I was able to build the project manually in the docker image used by travis, so the old version of GCC should not be a problem here

**Comment by lelit** [comment]
FYI, I just released version 1.7, using new cibuildwheel 1.0.0. Hopefully I will be able to inspect this PR in the following days.

**Comment by rdunklau** [comment]
I rebased the PR on master.

**Comment by lelit** [comment]
Great, thank you. If nothing distracts me, I will try this out tomorrow...

**Comment by lelit** [comment]
Ronan, am I supposed to do something particular after the initial clone?

Just to be sure, I cloned _your own_ branch in a new local tree and did a `make` to initialize the virtualenv. At that point, a `make build` gives me the following error:
```
make build
make: *** No rule to make target 'libpg_query/tmp/postgres/src/include/storage/lockdefs.h', needed by 'pglast/enums/lockdefs.py'.  Stop.
```

Effectively, the `libpg_query/tmp` tree is empty.
Then I tried to `cd libpg_query; make`, and that downloaded PG 12.1 sources, patched them and went on compiling it. It eventually produced `libpg_query.a`, but right after that, compiling `examples/simple_plpgsql.c`, failed with
```
cc -I. -I /home/lele/wip/pglast+PG12/libpg_query/tmp/postgres/src/pl/plpgsql/src/ -I /home/lele/wip/pglast+PG12/libpg_query/tmp/postgres/contrib/pgcrypto -I/home/lele/wip/pglast+PG12/libpg_query/tmp/postgres/src/include  -Wall -Wno-unused-function -Wno-unused-value -Wno-unused-variable -fno-strict-aliasing -fwrapv -fPIC -lm -ldl -lrt -pthread -flto -fvisibility=hidden -fdata-sections -ffunction-sections -g -Os -fno-inline -I. -o examples/simple_plpgsql -g examples/simple_plpgsql.c libpg_query.a -Wl,--gc-sections,--as-needed
/usr/bin/ld: /tmp/simple_plpgsql.Nyebwr.ltrans4.ltrans.o: in function `get_namespace_oid':
/home/lele/wip/pglast+PG12/libpg_query/namespace.c:3039: undefined reference to `GetSysCacheOid'
/usr/bin/ld: /tmp/simple_plpgsql.Nyebwr.ltrans4.ltrans.o: in function `LookupExplicitNamespace':
/home/lele/wip/pglast+PG12/libpg_query/namespace.c:2907: undefined reference to `pg_namespace_aclcheck'
/usr/bin/ld: /home/lele/wip/pglast+PG12/libpg_query/namespace.c:2909: undefined reference to `aclcheck_error'
/usr/bin/ld: /tmp/simple_plpgsql.Nyebwr.ltrans4.ltrans.o: in function `recomputeNamespacePath.lto_priv.0':
/home/lele/wip/pglast+PG12/libpg_query/namespace.c:3750: undefined reference to `SearchSysCache1'
/usr/bin/ld: /home/lele/wip/pglast+PG12/libpg_query/namespace.c:3757: undefined reference to `ReleaseSysCache'
/usr/bin/ld: /home/lele/wip/pglast+PG12/libpg_query/namespace.c:3788: undefined reference to `pg_namespace_aclcheck'
/usr/bin/ld: /tmp/simple_plpgsql.Nyebwr.ltrans4.ltrans.o: in function `plpgsql_yyparse':
/home/lele/wip/pglast+PG12/libpg_query/namespace.c:246: undefined reference to `get_database_name'
/usr/bin/ld: /home/lele/wip/pglast+PG12/libpg_query/namespace.c:321: undefined reference to `get_relname_relid'
/usr/bin/ld: /home/lele/wip/pglast+PG12/libpg_query/namespace.c:683: undefined reference to `get_relname_relid'
/usr/bin/ld: /tmp/simple_plpgsql.Nyebwr.ltrans4.ltrans.o: in function `plpgsql_yyparse':
/home/lele/wip/pglast+PG12/libpg_query/parse_type.c:137: undefined reference to `get_attnum'
/usr/bin/ld: /home/lele/wip/pglast+PG12/libpg_query/parse_type.c:151: undefined reference to `get_atttype'
/usr/bin/ld: /tmp/simple_plpgsql.Nyebwr.ltrans4.ltrans.o: in function `plpgsql_yyparse':
/home/lele/wip/pglast+PG12/libpg_query/namespace.c:2826: undefined reference to `get_database_name'
/usr/bin/ld: /tmp/simple_plpgsql.Nyebwr.ltrans4.ltrans.o: in function `plpgsql_yyparse':
/home/lele/wip/pglast+PG12/libpg_query/parse_type.c:182: undefined reference to `GetSysCacheOid'
/usr/bin/ld: /tmp/simple_plpgsql.Nyebwr.ltrans4.ltrans.o: in function `plpgsql_yyparse':
/home/lele/wip/pglast+PG12/libpg_query/namespace.c:790: undefined reference to `GetSysCacheOid'
/usr/bin/ld: /tmp/simple_plpgsql.Nyebwr.ltrans4.ltrans.o: in function `plpgsql_yyparse':
/home/lele/wip/pglast+PG12/libpg_query/parse_type.c:198: undefined reference to `get_array_type'
/usr/bin/ld: /home/lele/wip/pglast+PG12/libpg_query/parse_type.c:208: undefined reference to `SearchSysCache1'
/usr/bin/ld: /home/lele/wip/pglast+PG12/libpg_query/parse_type.c:418: undefined reference to `OidFunctionCall1Coll'
/usr/bin/ld: /home/lele/wip/pglast+PG12/libpg_query/parse_type.c:318: undefined reference to `ReleaseSysCache'
collect2: error: ld returned 1 exit status
```

For the record, I'm working in a Debian `sid`, where
```
$ gcc --version
gcc (Debian 9.2.1-21) 9.2.1 20191130
```

**Comment by lelit** [comment]
Forgot to say that even after that, a `make build` on the root directory fails with the same error, `libpg_query/tmp` is still empty.

**Comment by rdunklau** [comment]
I updated the Makefile for the missing target, sorry about that.
Compilation works for me if I perform the following steps:
```
git clone --recursive   git@github.com:rdunklau/pglast.git --branch PG12 /tmp/pglast12
cd /tmp/pglast12
make build
```
Is it possible you had some object files laying around after an attempt to build ?

**Comment by lelit** [comment]
No, that was a new fresh clone of your repo, but my attempt to manually build the libpg_query was wrong.

Anyway, I recloned your latest and now it builds without errors, but I end up with the following:
```
 python 
Python 3.7.5 (default, Oct 27 2019, 15:43:29) 
[GCC 9.2.1 20191022] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pglast
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/tmp/pglast+PG12/pglast/__init__.py", line 12, in <module>
    from .parser import get_postgresql_version, parse_plpgsql, parse_sql
ImportError: /tmp/pglast+PG12/pglast/parser.cpython-37m-x86_64-linux-gnu.so: undefined symbol: get_array_type
```

and indeed
```
$ nm libpg_query/libpg_query.a |grep get_array_type
         U get_array_type
         U get_array_type
         U get_array_type
         U get_array_type
         U get_array_type
         U get_array_type
nm: /tmp/pglast+PG12/libpg_query/tmp/postgres/src/backend/optimizer/geqo/geqo_misc.o: no symbols
nm: /tmp/pglast+PG12/libpg_query/tmp/postgres/src/backend/optimizer/geqo/geqo_mutation.o: no symbols
nm: /tmp/pglast+PG12/libpg_query/tmp/postgres/src/backend/optimizer/geqo/geqo_pmx.o: no symbols
nm: /tmp/pglast+PG12/libpg_query/tmp/postgres/src/backend/optimizer/geqo/geqo_cx.o: no symbols
nm: /tmp/pglast+PG12/libpg_query/tmp/postgres/src/backend/optimizer/geqo/geqo_px.o: no symbols
nm: /tmp/pglast+PG12/libpg_query/tmp/postgres/src/backend/optimizer/geqo/geqo_ox1.o: no symbols
nm: /tmp/pglast+PG12/libpg_query/tmp/postgres/src/backend/optimizer/geqo/geqo_ox2.o: no symbols
         U get_array_type
nm: /tmp/pglast+PG12/libpg_query/tmp/postgres/src/backend/port/atomics.o: no symbols
nm: /tmp/pglast+PG12/libpg_query/tmp/postgres/src/backend/utils/mmgr/memdebug.o: no symbols
```

**Comment by rdunklau** [comment]
Ok, I'll take a look at it with my debian-using colleague.

**Comment by lelit** [comment]
OOC, on what system do you develop usually?

**Comment by rdunklau** [comment]
Archlinux. My colleague successfully developed and tested the libpg_query port on debian, and I was able to build and at least parse a query on the old ubuntu docker image used by travis, so there shouldn't be a problem unless I introduced a problem recently :)

**Comment by PierreDucroquet** [comment]
Tsss @rdunklau always trying to blame others... :)
@lelit the dev was done on my debian testing and a bit on my debian sid machine. And I just tested git clone ; make ; make build and it worked (well, it did parse "SELECT 1+1", let's say I am pleased with simple things). GCC 9.2.1-21, same as you.

Since I am the broken mind behind this silly idea, I'll dig further. I think we did not foresee that there may be things "taken" from the env in the ./configure of postgresql, leading to different code paths being used.

FYI, the error is not that the symbol is undefined, the error is that this symbol is being used. On my build, `strings parser.*.so | grep get_array_type` returns nothing. Could you tell which symbol in your .so file refers to get_array_type? (I tend not to be very subtle, so I suggest `objdump -S parser.*.so | less` :) )

**Comment by PierreDucroquet** [comment]
@lelit also, much more important than the compiler in this scenario: which linker did you use? I use GNU ld 2.33.1

**Comment by lelit** [comment]
Yes, of course, my check on `nm libpg_query/libpg_query.a |grep get_array_type` was indeed superfluous :smiley:

Wrt `objdump -S`, I'm not sure what I should report, it goes like this:
```
$ objdump -S parser.cpython-37m-x86_64-linux-gnu.so | grep -C 5 get_array_type
00000000000080d0 <memchr@plt>:
    80d0:	ff 25 92 6f 12 00    	jmpq   *0x126f92(%rip)        # 12f068 <memchr@GLIBC_2.2.5>
    80d6:	68 0a 00 00 00       	pushq  $0xa
    80db:	e9 40 ff ff ff       	jmpq   8020 <.plt>

00000000000080e0 <get_array_type@plt>:
    80e0:	ff 25 8a 6f 12 00    	jmpq   *0x126f8a(%rip)        # 12f070 <get_array_type>
    80e6:	68 0b 00 00 00       	pushq  $0xb
    80eb:	e9 30 ff ff ff       	jmpq   8020 <.plt>

00000000000080f0 <get_atttype@plt>:
    80f0:	ff 25 82 6f 12 00    	jmpq   *0x126f82(%rip)        # 12f078 <get_atttype>
--
   2ab68:	eb c9                	jmp    2ab33 <plpgsql_yyparse+0x16f2>
   2ab6a:	45 31 ff             	xor    %r15d,%r15d
   2ab6d:	49 83 7d 28 00       	cmpq   $0x0,0x28(%r13)
   2ab72:	74 0b                	je     2ab7f <plpgsql_yyparse+0x173e>
   2ab74:	44 89 ff             	mov    %r15d,%edi
   2ab77:	e8 64 d5 fd ff       	callq  80e0 <get_array_type@plt>
   2ab7c:	41 89 c7             	mov    %eax,%r15d
   2ab7f:	45 85 ff             	test   %r15d,%r15d
   2ab82:	75 50                	jne    2abd4 <plpgsql_yyparse+0x1793>
   2ab84:	45 31 c0             	xor    %r8d,%r8d
   2ab87:	48 8d 0d 3a 8f 0e 00 	lea    0xe8f3a(%rip),%rcx        # 113ac8 <__func__.19955>
```
Please tell me whatever detail may help!

Finally:
```
$ ld --version
GNU ld (GNU Binutils for Debian) 2.33.1
```

**Comment by PierreDucroquet** [comment]
This symbol should not be included. In my case it was not.
I can give you the list of symbols I have on my side, 1388 symbols.

[symbols.txt](https://github.com/lelit/pglast/files/3918675/symbols.txt)

**Comment by lelit** [comment]
Yay, I found my issue: I was missing `bison`, and in that case, at some point in the PG build there's an alert message about that, but I didn't notice it until this morning, when I followed the log more closely.
So, great, this compiles without issues and all tests pass!!
Thanks for the patience.

**Comment by lelit** [comment]
Could the CI error be related? I fail to see, in the CI log, where it compiles the underlying libpg_query...

**Comment by lelit** [comment]
No, apparently `bison` is present in the `quay.io/pypa/manylinux2010_x86_64` image used by `cibuildwheel`.

File: setup.py [hunk 1]
```diff
@@ -30,7 +30,11 @@
 
 
 LIBPG_QUERY_DIR = str(here / 'libpg_query')
+INCLUDE_DIR = str(here / 'libpg_query' / 'tmp' / 'postgres' / 'src' /
+                  'include')
 
+CFLAGS="-flto -fvisibility=hidden -fdata-sections -ffunction-sections"
+LDFLAGS="-Wl,--gc-sections,--as-needed"
```
Comment (lelit) [inline setup.py]: @PierreDucroquet, @rdunklau: shouldn't these be passed to the `Extension` below as respectively `extra_compile_args` and `extra_link_args`?

Comment (rdunklau) [reply setup.py]: You're right, except we need to pass all the flags as compile and link args. I pushed the fix.

**Comment by lelit** [comment]
Another thing that should be checked:
```
$ python
Python 3.7.5 (default, Oct 27 2019, 15:43:29) 
[GCC 9.2.1 20191022] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import pglast
>>> pglast.get_postgresql_version()
(12, 0, 1)
```

I fail to see how is that injected in the `libpg_query`: commit 223f7a0986a1b8646a9c44d6c0d06514b70c987a removed the definitions from `pg_query.h` (maybe also the `// Postgres version information` should be removed?), so, where does that 12.0.1 come from now?

**Comment by rdunklau** [comment]
You're right, this should be removed as well as the no-ops in libpg_query makefile.
The PG_VERSION_NUM comes from pg_config.h, included by c.h, included by pg_query.h.

**Comment by rdunklau** [comment]
Thanks to Pierre's fix on our libpg_query branch, pglast successfully build using cibuildwheel. However, I don't understand why a test is not passing in the CI whereas it does on my local machine...  I tested using python 3.8 on my machine.

**Comment by lelit** [comment]
Great!

The "?" placeholder handling is a libpg_query tweak to standard parser... maybe, for whatever reason, that patch is not applied within the CI environment?

**Comment by lelit** [comment]
@rdunklau, trying to understand the issue I have with the `PG_VERSION_NUM` I did a pull from your pglast branch, but I got:
```
$ git fetch
remote: Enumerating objects: 13, done.
remote: Counting objects: 100% (13/13), done.
remote: Compressing objects: 100% (9/9), done.
remote: Total 10 (delta 6), reused 5 (delta 1), pack-reused 0
Decompressione degli oggetti in corso: 100% (10/10), fatto.
Da https://github.com/rdunklau/pglast
   51a5c1a..efa9a35  PG12       -> origin/PG12
Fetching submodule libpg_query
error: Il server non consente richieste per l'oggetto non pubblicizzato b2a1461214621ede050fb62d1f84fa6e467daa2c
```
(sorry for the italian language) by any chance you didn't push latest commits on libpg_query?

**Comment by lelit** [comment]
Uhm, ignore that: doing a fresh clone works... so something's wrong on my tree.

**Comment by lelit** [comment]
See issue #38, the problem with `PG_VERSION_NUM` is definitely not related with this PR.

**Comment by lelit** [comment]
FYI, I merged this into the [v2 branch](https://github.com/lelit/pglast/tree/v2).

**Comment by rdunklau** [comment]
Since it's been merged in the v2 branch, I'm closing this PR.

**Pull request closed without merging**